adapta/storage/query_enabled/_models.py

```diff
@@ -0,0 +1,146 @@
+from typing import TypeVar, Generic, Type, Iterator, Union, final, Optional
+
+import pandas
+
+from adapta.storage.models.base import DataPath
+from adapta.storage.models.filter_expression import Expression
+
+TCredential = TypeVar("TCredential")  # pylint: disable=C0103
+TSettings = TypeVar("TSettings")  # pylint: disable=C0103
+
+CONNECTION_STRING_REGEX = r"^qes:\/\/class=(.*?);plaintext_credentials=(.*?);settings=(.*?)$"
+
+
+@final
+class BundledQes(Enum):
+    """
+    QES Implementations aliases that are bundled with Adapta.
+    """
+
+    DELTA = "adapta.storage.query_enabled.DeltaQes"
+    ASTRA = "adapta.storage.query_enabled.AstraQes"
+
+
+BUNDLED_STORES = {store.name: store.value for store in BundledQes}
+
+
+class QueryEnabledStore(Generic[TCredential, TSettings], ABC):
+    """
+    QES base class.
+    """
+
+    def __init__(self, credentials: TCredential, settings: TSettings):
+        self._credentials = credentials
+        self._settings = settings
+
+    @property
+    def credentials(self) -> TCredential:
+        """
+        Returns the credentials for this store type.
+        """
+        return self._credentials
+
+    @property
+    def settings(self) -> TSettings:
+        """
+        Returns the address to connect to, if applicable.
+        """
+        return self._settings
+
+    def open(self, path: DataPath) -> "QueryEnabledStoreReader":
+        """
+        Construct a reader object for QES to proxy to the underlying store implementation.
+        """
+        return QueryEnabledStoreReader(self, path)
+
+    @abstractmethod
+    def _apply_filter(
+        self, path: DataPath, filter_expression: Expression, columns: list[str]
+    ) -> Union[pandas.DataFrame, Iterator[pandas.DataFrame]]:
+        """
+        Applies the provided filter expression to this Store and returns the result in a pandas DataFrame
+        """
+
+    @abstractmethod
+    def _apply_query(self, query: str) -> Union[pandas.DataFrame, Iterator[pandas.DataFrame]]:
+        """
+        Applies a plaintext query to this Store and returns the result in a pandas DataFrame
+        """
+
+    @classmethod
+    @abstractmethod
+    def _from_connection_string(cls, connection_string: str) -> "QueryEnabledStore[TCredential, TSettings]":
+        """
+        Constructs the connection from a connection string
+        """
+
+    @staticmethod
+    def from_string(connection_string: str) -> "QueryEnabledStore[TCredential, TSettings]":
+        """
+        Constructs a concrete QES instance from a connection string.
+        """
+
+        def get_qes_class(name: str) -> Type[QueryEnabledStore[TCredential, TSettings]]:
+            return locate(BUNDLED_STORES.get(name, name))
+
+        class_name, _, _ = re.findall(re.compile(CONNECTION_STRING_REGEX), connection_string)[0]
+        class_object = get_qes_class(class_name)
+        return class_object._from_connection_string(connection_string)
+
+
+@final
+class QueryEnabledStoreReader:
```

Comment:
It actually does not do any reading. It just builds stuff. `QueryBuilder`?

Reply:
> QueryConfigurationBuilder